rfcs/text/0012_background_sessions.md

```diff
@@ -0,0 +1,146 @@
+
+# Detailed design
+
+From the perspective of a user, when they run a view that takes a long time (A dashboard, a discover search, a SIEM timeline, etc.), they will recieve a notification stating that it can be sent to background, and returned to later. If they choose to do so, that view will be treated as a "Background Session": It will be stored and tracked for progress, allowing the user to come back and retrive the results in the future.
+
+## Session Management
+
+A session is a grouping of search requests, who's results are required to produce a specific view of Kibana. 
+It's an application's responsibility to manage its' sessions, by using the `data.search.session` APIs. 
+
+### Starting a new session
+
+To start a session, an application should call `const sessionId = data.search.session.start();`, where `data` is the start contract of the data plugin.
+
+If the application uses the services provided by the `data.search` plugin, it only needs to pass down that `sessionId` to the search API, and the `search` service will track and restore requests internally.
+
+ - If the request is issued using the `data.search.search`, use: 
+ ```
+    data.search.search({
+        params,
+        sessionId,
+    }
+ ```
+ - If the request is issued using `SearchSource`, use `searchSource.setSessionId(sessionId)`, before calling `fetch`.
+
+If the application doesn't use `data.search` services, you may still associate your requests with a `BackgroundSession`, by using the 'backgroundSession` service directly. Examples will be provided in the Detailed Design section.
+
+### Closing a session
+
+Each session is automatically closed when a new one is started. You may also call `data.search.session.clear()` to close a session without starting a new one.
+
+Sessions are also closed automatically when navigating between applications.
+
+If there is no open session, Kibana will act as before, suggesting a user to "Run Beyond Timeout" instead of "Run in Background".
+
+### Restoring a session
+
+Restoring a session means loading the results previously generated for each request within the session.  
+
+If, for example, the `sessionId` to be 
+
+Doing so requires a `sessionId` and *calling each request with the exact same parameters*. If the request parameters and the `sessionId` match, the stored results will be returned immediatelly, instead of re-running the search.
+
+If the `sessionId` is not found, the request parameters don't match or the stored results are expired, the requests will be re-run. 
+
+### Embeddable integration
+
+TBD 
+
+## Background Session Service
+
+The main component of this RFC is the background session service.
+
+It is a server side service, that tracks and stores sessions and all associated requests. 
+When a user starts executing a session, the service stores information in memory, until the user decides to "Send to Background" or until the expiration time.
```

Comment:
Will this happen over multiple HTTP requests? When Kibana is deployed behind a load-balancer in a highly-available deployment, multiple HTTP requests will likely be served by multiple different instances of Kibana with different "memory".

Reply:
There will be one saved object, but multiple servers may be updating the saved object (depending on where search requests are routed).